client/executor/wasmtime/src/instance_wrapper.rs

```diff
@@ -0,0 +1,245 @@
+		let entrypoint = export
+			.func()
+			.ok_or_else(|| Error::from(format!("Export {} is not a function", name)))?;
+		match (entrypoint.ty().params(), entrypoint.ty().results()) {
+			(&[wasmtime::ValType::I32, wasmtime::ValType::I32], &[wasmtime::ValType::I64]) => {}
+			_ => {
+				return Err(Error::from(format!(
+					"method {} have an unsupported signature",
+					name
+				)))
+			}
+		}
+		Ok(entrypoint.clone())
+	}
+
+	/// Returns an indirect function table of this instance.
+	pub fn table(&self) -> Option<&Table> {
+		self.table.as_ref()
+	}
+
+	/// Returns the byte size of the linear memory instance attached to this instance.
+	pub fn memory_size(&self) -> u32 {
+		self.memory.data_size() as u32
+	}
+
+	/// Reads `__heap_base: i32` global variable and returns it.
+	///
+	/// If it doesn't exist, not a global or of not i32 type returns an error.
+	pub fn extract_heap_base(&self) -> Result<u32> {
+		let heap_base_export = self
+			.instance
+			.get_export("__heap_base")
+			.ok_or_else(|| Error::from("__heap_base is not found"))?;
+
+		let heap_base_global = heap_base_export
+			.global()
+			.ok_or_else(|| Error::from("__heap_base is not a global"))?;
+
+		let heap_base = heap_base_global
+			.get()
+			.i32()
+			.ok_or_else(|| Error::from("__heap_base is not a i32"))?;
+
+		Ok(heap_base as u32)
+	}
+}
+
+fn get_table(instance: &Instance) -> Option<Table> {
+	instance
+		.get_export("__indirect_function_table")
+		.and_then(|export| export.table())
+		.cloned()
+}
+
+fn get_linear_memory(instance: &Instance) -> Result<Memory> {
+	let memory_export = instance
+		.get_export("memory")
+		.ok_or_else(|| Error::from("memory is not exported under `memory` name"))?;
+
+	let memory = memory_export
+		.memory()
+		.ok_or_else(|| Error::from("the `memory` export should have memory type"))?
+		.clone();
+
+	Ok(memory)
+}
+
+/// Functions realted to memory.
+impl InstanceWrapper {
+	/// Read data from a slice of memory into a destination buffer.
+	///
+	/// Returns an error if the read would go out of the memory bounds.
+	pub fn read_memory_into(&self, address: Pointer<u8>, dest: &mut [u8]) -> Result<()> {
+		unsafe {
+			// This should be safe since we don't grow up memory while caching this reference and
+			// we give up the reference before returning from this function.
+			let memory = self.memory_as_slice();
+
+			let range = util::checked_range(address.into(), dest.len(), memory.len())
+				.ok_or_else(|| Error::Other("memory read is out of bounds".into()))?;
+			dest.copy_from_slice(&memory[range]);
+			Ok(())
+		}
+	}
+
+	/// Write data to a slice of memory.
+	///
+	/// Returns an error if the write would go out of the memory bounds.
+	pub fn write_memory_from(&self, address: Pointer<u8>, data: &[u8]) -> Result<()> {
+		unsafe {
+			// This should be safe since we don't grow up memory while caching this reference and
+			// we give up the reference before returning from this function.
+			let memory = self.memory_as_slice_mut();
```

Comment:
We need to ensure that `InstanceWrapper` is `!Sync` for this to be safe.

Reply:
yeah i think it is worth specifying explicitly.